test/fixtures/azure-fixture/src/main/java/fixture/azure/MockAzureBlobStore.java

```diff
@@ -0,0 +1,487 @@
+
+        protected final Object writeLock = new Object();
+        protected final Lease lease = new Lease();
+
+        @Override
+        public String acquireLease(String proposedLeaseId, int leaseTimeSeconds) {
+            synchronized (writeLock) {
+                return lease.acquire(proposedLeaseId, leaseTimeSeconds);
+            }
+        }
+
+        @Override
+        public void releaseLease(String leaseId) {
+            synchronized (writeLock) {
+                lease.release(leaseId);
+            }
+        }
+
+        @Override
+        public void assertCanRead(String leaseId) {
+            lease.assertCanRead(leaseId);
+        }
+
+        @Override
+        public void assertCanWrite(String leaseId) {
+            lease.assertCanWrite(leaseId);
+        }
+    }
+
+    private static class MockAzureBlockBlob extends AbstractAzureBlob implements AzureBlob {
+        private final Map<String, BytesReference> blocks;
+        private BytesReference contents = EMPTY;
+
+        private MockAzureBlockBlob() {
+            this.blocks = new ConcurrentHashMap<>();
+        }
+
+        public void putBlock(String blockId, BytesReference content, String leaseId) {
+            synchronized (writeLock) {
+                lease.assertCanWrite(leaseId);
+                this.blocks.put(blockId, content);
+            }
+        }
+
+        public void putBlockList(List<String> blockIds, String leaseId) throws BadRequestException {
+            synchronized (writeLock) {
+                lease.assertCanWrite(leaseId);
+                final List<String> unresolvedBlocks = blockIds.stream().filter(bId -> blocks.containsKey(bId) == false).toList();
+                if (unresolvedBlocks.isEmpty() == false) {
+                    logger.warn("Block list contained non-existent block IDs: {}", unresolvedBlocks);
+                    throw new BadRequestException("InvalidBlockList", "The specified blocklist is invalid.");
+                }
+                final BytesReference[] resolvedContents = blockIds.stream().map(blocks::get).toList().toArray(new BytesReference[0]);
+                contents = CompositeBytesReference.of(resolvedContents);
+            }
+        }
+
+        private boolean matches(String ifNoneMatchHeaderValue) {
+            if (ifNoneMatchHeaderValue == null) {
+                return false;
+            }
+            // We only support *
+            if ("*".equals(ifNoneMatchHeaderValue)) {
+                return true;
+            }
+            throw new BadRequestException("UnsupportedHeader", "The test fixture only supports * for if-none-match");
+        }
+
+        @Override
+        public synchronized void setContents(BytesReference contents, String leaseId) {
+            synchronized (writeLock) {
+                lease.assertCanWrite(leaseId);
+                this.contents = contents;
+                this.blocks.clear();
+            }
+        }
```

Comment:
In general I made the lease check and subsequent write atomic. I couldn't see any indication of whether this is the case in Azure, but it should be AT LEAST as safe as Azure.

Also lease acquires, lease releases and blob updates are mutually exclusive

Reply:
Makes sense to me.

> it should be AT LEAST as safe as Azure.

Note that really we're generally trying to be as unsafe as possible here while still staying within the documented behaviour, i.e. the goal is really to be AT MOST as safe as Azure. If there's ways that things can legitimately fail or behave weirdly then we should be trying to do that here to make sure that we behave correctly in tests even in the face of anomalies that might be possible yet vanishingly rare against the real service.

Reply:
Yep, understood. I think what I've got is probably about minimal for the guarantee, i.e. probably want to prevent

```
C1: send update | check lease (OK) |             |                  |              | apply update | respond OK |
C2:             | acquire lease    | send update | check lease (OK) | apply update | respond OK   |
```

Which while improbable, would be possible if lease-check and update were not atomic